src/pcp.rs

```diff
@@ -0,0 +1,596 @@
+
+    // Run the validity crciut with a "shim" gadget that records the value of each input wire of
+    // each gadget evaluation.
+    let mut shim = ProveShimGadget::new(&mut g, g_calls);
+    let _ = x.valid(&mut shim, &rand_vec(x.valid_rand_len()));
+
+    // Construct the intermediate proof polynomials `f[0], ..., f[l-1]`. Also, record in the slice
+    // `data[...l]` the value of `f[i](1)`, i.e., the first point at which polynomial `f[i]` was
+    // interpolated.
+    let mut f = vec![vec![F::zero(); m]; l];
+    let m_inv = F::from(F::Integer::try_from(m).unwrap()).inv();
+    for i in 0..l {
+        data[i] = shim.f_vals[i][0];
+        discrete_fourier_transform(&mut f[i], &shim.f_vals[i], m)?;
+        discrete_fourier_transform_inv_finish(&mut f[i], m, m_inv);
+    }
+
+    // Construct the proof polynomial `data[l..] = G(f[0], ..., f[l-1])`.
+    g.call_poly(&mut data[l..], &f)?;
+
+    let poly_len = poly_deg(&data[l..]);
+    data.truncate(l + poly_len + 1);
+    Ok(Proof { data })
+}
+
+// A "shim" gadget used during proof generation to record the points at which the intermediate
+// proof polynomials are interpolated.
+struct ProveShimGadget<'a, F, G>
+where
+    F: FieldElement,
+    G: Gadget<F>,
+{
+    inner: &'a mut G,
+    f_vals: Vec<Vec<F>>, // Points at which intermediate proof polynmoials are interpolated.
+    ct: usize,
+}
+
+impl<'a, F, G> ProveShimGadget<'a, F, G>
+where
+    F: FieldElement,
+    G: Gadget<F>,
+{
+    fn new(inner: &'a mut G, gadget_calls: usize) -> Self {
+        let mut f_vals = vec![vec![F::zero(); gadget_calls + 1]; inner.call_in_len()];
+        for i in 0..f_vals.len() {
+            // Choose a random field element as first point on the i-th proof polynomial.
+            f_vals[i][0] = F::rand();
+        }
+        Self {
+            inner,
+            f_vals,
+            ct: 1,
+        }
+    }
+}
+
+impl<'a, F, G> GadgetCallOnly<F> for ProveShimGadget<'a, F, G>
+where
+    F: FieldElement,
+    G: Gadget<F>,
+{
+    fn call(&mut self, inp: &[F]) -> Result<F, PcpError> {
+        for i in 0..inp.len() {
+            self.f_vals[i][self.ct] = inp[i];
+        }
+        self.ct += 1;
+        self.inner.call(inp)
+    }
+
+    fn call_in_len(&self) -> usize {
+        self.inner.call_in_len()
+    }
+}
+
+/// The output of `prove`, a proof of an input's validity.
+#[derive(Clone, Debug)]
+pub struct Proof<F: FieldElement> {
+    data: Vec<F>,
+}
+
+impl<F: FieldElement> Proof<F> {
+    /// Returns a reference to the underlying data.
+    pub fn as_slice(&self) -> &[F] {
+        &self.data
+    }
+
+    /// Constructs a proof from the provided data. This call takes ownership of `data`.
+    pub fn from(data: Vec<F>) -> Self {
+        Self { data }
+    }
+}
+
+/// Generate the verification message for an input and proof (or input share and proof share).
+pub fn query<F, G, T>(x: &T, pf: &Proof<F>, rand: &[F]) -> Result<Verifier<F>, PcpError>
```

Comment:
I kinda feel like this should be a method, but it's hard to say if it should be a method on `Type` or on `Proof`. I think maybe what's called for here is `Verifier<F: FieldElement, T: Type>::new(x: &T, pf: &Proof<F>, rand: &[F]) -> Result<Verifier<F>, PcpError>`.

Reply:
I don't like this suggestion because it breaks the nice correspondence between `prove`, `query`, and `decide` hand the description of the fully linear PCP in the paper, which literally names these three algorithms. Right now we have
```rust
let pf = prove(&x).unwrap();
let vf = query(&x, &pf, &rand_vec(x.valid_rand_len()).unwrap();
let res = decide(&x, &vf).unwrap();
```
and with your suggestions we would do (something like)
```rust
let pf = x.prove().unwrap();
let vf = Verifier::new(&x, &pf, &rand_vec(x.valid_rand_len()).unwrap();
let = decide(&x, &vf).unwrap();
```
This doesn't correspond nearly as nicely with the paper.

By the way, this is purely a matter of taste. I don't have any other reason, other than I think it makes it easier to read the code.

Reply:
I was ambivalent about it too, and I think the argument about corresponding to the paper's definitions is good enough to tip the balance towards the free-function formulation you have here.